File: captainhcg/go/111-minimum-depth-of-binary-tree.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func minDepthHelper(small, large int) int {
    // small and large cannot both be 0
    if small == 0 {
        return large + 1
    }
    return small + 1
}

func minDepth(root *TreeNode) int {
    if root == nil {
        return 0
    }
    left, right := minDepth(root.Right), minDepth(root.Left)
    if left > right {
        return minDepthHelper(right, left)
    } else {
        return minDepthHelper(left, right)
    }
}

File: captainhcg/py/111-minimum-depth-of-binary-tree.py
class Solution(object):
    def minDepth(self, root):
        """
        :type root: TreeNode
        :rtype: int
        """
        if not root:
            return 0
        if not root.left and not root.right:
            return 1
        if root.left and root.right:
            return min(self.minDepth(root.left), self.minDepth(root.right)) + 1
        return self.minDepth(root.left or root.right) + 1
